### PR TITLE
Error Message Enhancement

### PR DESCRIPTION
**Pull request by zhhsplendid**
This PR enhances error messages of several API/OPs:

ParallelExecutor (python && C++)
Executor (python && C++)
StaticRNN (python)
IfElse (python)
cond (python)
split_lod_tensor (python && C++)

### CODE REVIEW

File: paddle/fluid/framework/executor.cc [hunk 1]
```diff
@@ -367,14 +389,20 @@ std::vector<std::shared_ptr<ExecutorPrepareContext>> Executor::Prepare(
...
+      platform::errors::InvalidArgument("skip_ref_cnt_vars should be either "
+                                        "empty or equals to block number %d",
+                                        block_ids.size()));
   std::vector<std::shared_ptr<ExecutorPrepareContext>> result;
   size_t idx = 0;
   for (auto& bid : block_ids) {
-    PADDLE_ENFORCE_LT(static_cast<size_t>(bid), program.Size());
+    PADDLE_ENFORCE_LT(static_cast<size_t>(bid), program.Size(),
+                      platform::errors::InvalidArgument(
+                          "Input block id = %d, but it should be less than "
```
Comment (wzzju) [inline paddle/fluid/framework/executor.cc]: `%zu` may be more suitable here.

File: paddle/fluid/operators/split_lod_tensor_op.cc [hunk 2]
```diff
@@ -146,34 +147,39 @@ class SplitLoDTensorOpProtoMaker : public framework::OpProtoAndCheckerMaker {
...
+                          "SplitLoDTensorOp must have input X."));
+    PADDLE_ENFORCE_EQ(context->HasInput("Mask"), true,
+                      platform::errors::InvalidArgument(
+                          "SplitLoDTensorOp must have input Mask."));
+    PADDLE_ENFORCE_EQ(context->HasOutput("OutTrue"), true,
+                      platform::errors::InvalidArgument(
+                          "SplitLoDTensorOp must have output OutTrue."));
+    PADDLE_ENFORCE_EQ(context->HasOutput("OutFalse"), true,
+                      platform::errors::InvalidArgument(
+                          "SplitLoDTensorOp must have output OutFalse."));
```
Comment (wzzju) [inline paddle/fluid/operators/split_lod_tensor_op.cc]: Please use`OP_INOUT_CHECK `instead.

Comment (zhhsplendid) [reply paddle/fluid/operators/split_lod_tensor_op.cc]: Done

File: paddle/fluid/framework/executor.cc (hunk 1 above)
Comment (zhhsplendid) [reply paddle/fluid/framework/executor.cc]: Done

File: paddle/fluid/framework/executor.cc [hunk 3]
```diff
@@ -203,29 +206,36 @@ static bool has_feed_operators(
...
+                            feed_target_name));
     }
   }
 
   if (feed_count > 0) {
     PADDLE_ENFORCE_EQ(
         feed_count, feed_targets.size(),
-        "The number of feed operators should match 'feed_targets'");
+        platform::errors::PreconditionNotMet(
+            "The number of feed operators should match 'feed_targets'"));
```
Comment (Aurelius84) [inline paddle/fluid/framework/executor.cc]: `"The number of feed operators should match 'feed_targets, but received feed_count: %d, required feed_targets.size():  %d.",  feed_count, feed_targets.size()` ?

Is it better to print`feed_count` and ` feed_targets.size()` ?

File: paddle/fluid/framework/executor.cc [hunk 4]
```diff
@@ -247,29 +257,37 @@ static bool has_fetch_operators(
...
+                            fetch_target_name));
     }
   }
 
   if (fetch_count > 0) {
     PADDLE_ENFORCE_EQ(
         fetch_count, fetch_targets.size(),
-        "The number of fetch operators should match 'fetch_targets'");
+        platform::errors::PreconditionNotMet(
+            "The number of fetch operators should match 'fetch_targets'"));
```
Comment (Aurelius84) [inline paddle/fluid/framework/executor.cc]: Same as above.

File: paddle/fluid/framework/parallel_executor.cc [hunk 5]
```diff
@@ -216,24 +224,29 @@ class ParallelExecutorPrivate {
...
-                     "inter_nranks:%d < 1",
-                     bst->hierarchical_allreduce_inter_nranks_);
-      PADDLE_ENFORCE(
-          (bst->num_trainers_ % bst->hierarchical_allreduce_inter_nranks_ == 0),
-          "num_trainers:%llu mod inter_nranks:%d != 0", bst->num_trainers_,
-          bst->hierarchical_allreduce_inter_nranks_);
+      PADDLE_ENFORCE_GT(bst->num_trainers_, 1,
+                        platform::errors::PreconditionNotMet(
+                            "num_trainers is %llu, which is less than 1",
+                            bst->num_trainers_));
```
Comment (Aurelius84) [inline paddle/fluid/framework/parallel_executor.cc]: `"The num_trainers should be greater than 1, but received %llu.", bst->num_trainers_` is better?

File: paddle/fluid/framework/parallel_executor.cc [hunk 6]
```diff
@@ -216,24 +224,29 @@ class ParallelExecutorPrivate {
...
+      PADDLE_ENFORCE_GT(bst->hierarchical_allreduce_inter_nranks_, 1,
+                        platform::errors::PreconditionNotMet(
+                            "inter_nranks is %d, which is less than 1",
+                            bst->hierarchical_allreduce_inter_nranks_));
```
Comment (Aurelius84) [inline paddle/fluid/framework/parallel_executor.cc]: `"The inter_nranks should be greater than 1, but received %d.", bst->hierarchical_allreduce_inter_nranks_`

File: paddle/fluid/framework/executor.cc (hunk 3 above)
Comment (zhhsplendid) [reply paddle/fluid/framework/executor.cc]: Done

File: paddle/fluid/framework/executor.cc (hunk 4 above)
Comment (zhhsplendid) [reply paddle/fluid/framework/executor.cc]: Done

File: paddle/fluid/framework/parallel_executor.cc (hunk 5 above)
Comment (zhhsplendid) [reply paddle/fluid/framework/parallel_executor.cc]: Done

File: paddle/fluid/framework/parallel_executor.cc (hunk 6 above)
Comment (zhhsplendid) [reply paddle/fluid/framework/parallel_executor.cc]: Done

**Review by Aurelius84 — DISMISSED** [review]
LGTM

File: paddle/fluid/framework/executor.cc [hunk 7]
```diff
@@ -203,29 +206,38 @@ static bool has_feed_operators(
     if (op->Type() == kFeedOpType) {
       feed_count++;
       // The input variable's name of feed_op should be feed_holder_name.
-      PADDLE_ENFORCE_EQ(op->Input("X")[0], feed_holder_name,
-                        "Input to feed op should be '%s'", feed_holder_name);
+      PADDLE_ENFORCE_EQ(
+          op->Input("X")[0], feed_holder_name,
+          platform::errors::PreconditionNotMet(
+              "Input to feed op should be '%s'", feed_holder_name));
```
Comment (liym27) [inline paddle/fluid/framework/executor.cc]: May be better to add message "but received XX".

File: paddle/fluid/framework/executor.cc [hunk 8]
```diff
@@ -203,29 +206,38 @@ static bool has_feed_operators(
...
+      PADDLE_ENFORCE_NOT_NULL(
+          var,
+          platform::errors::PreconditionNotMet(
+              "Block should already have a '%s' variable", feed_holder_name));
       PADDLE_ENFORCE_EQ(var->GetType(), proto::VarType::FEED_MINIBATCH,
-                        "'%s' variable should be 'FEED_MINIBATCH' type",
-                        feed_holder_name);
+                        platform::errors::PreconditionNotMet(
+                            "'%s' variable should be 'FEED_MINIBATCH' type",
+                            feed_holder_name));
```
Comment (liym27) [inline paddle/fluid/framework/executor.cc]: May be better to add message "but received XX".

File: paddle/fluid/framework/executor.cc [hunk 9]
```diff
@@ -247,29 +259,39 @@ static bool has_fetch_operators(
     if (op->Type() == kFetchOpType) {
       fetch_count++;
       // The output variable's name of fetch_op should be fetch_holder_name.
-      PADDLE_ENFORCE_EQ(op->Output("Out")[0], fetch_holder_name,
-                        "Output of fetch op should be '%s'", fetch_holder_name);
+      PADDLE_ENFORCE_EQ(
+          op->Output("Out")[0], fetch_holder_name,
+          platform::errors::PreconditionNotMet(
+              "Output of fetch op should be '%s'", fetch_holder_name));
```
Comment (liym27) [inline paddle/fluid/framework/executor.cc]: Same as above

File: paddle/fluid/framework/parallel_executor.cc [hunk 10]
```diff
@@ -178,7 +180,9 @@ class ParallelExecutorPrivate {
     for (int i = 0; i < static_cast<int>(bst.nccl_comm_num_); i++) {
       std::string var_name = platform::GetFlatNCCLVarName(i);
       auto nccl_id_var = scope->FindVar(var_name);
-      PADDLE_ENFORCE(nccl_id_var, "can't find %s nccl_id_var", var_name);
+      PADDLE_ENFORCE_NOT_NULL(
+          nccl_id_var,
+          platform::errors::NotFound("can't find %s nccl_id_var", var_name));
```
Comment (liym27) [inline paddle/fluid/framework/parallel_executor.cc]: Please capitalize the error message.

File: paddle/fluid/framework/parallel_executor.cc [hunk 11]
```diff
@@ -191,7 +195,9 @@ class ParallelExecutorPrivate {
       for (int i = 0; i < static_cast<int>(bst.nccl_comm_num_); i++) {
         std::string var_name = platform::GetHierarchicalInterNCCLVarName(i);
         auto nccl_id_var = scope->FindVar(var_name);
-        PADDLE_ENFORCE(nccl_id_var, "can't find %s nccl_id_var", var_name);
+        PADDLE_ENFORCE_NOT_NULL(
+            nccl_id_var,
+            platform::errors::NotFound("can't find %s nccl_id_var", var_name));
```
Comment (liym27) [inline paddle/fluid/framework/parallel_executor.cc]: Same as above.

File: paddle/fluid/framework/parallel_executor.cc [hunk 12]
```diff
@@ -200,7 +206,9 @@ class ParallelExecutorPrivate {
       for (int i = 0; i < static_cast<int>(bst.nccl_comm_num_); i++) {
         std::string var_name = platform::GetHierarchicalExterNCCLVarName(i);
         auto nccl_id_var = scope->FindVar(var_name);
-        PADDLE_ENFORCE(nccl_id_var, "can't find %s nccl_id_var", var_name);
+        PADDLE_ENFORCE_NOT_NULL(
+            nccl_id_var,
+            platform::errors::NotFound("can't find %s nccl_id_var", var_name));
```
Comment (liym27) [inline paddle/fluid/framework/parallel_executor.cc]: Same as above.

**Review by liym27 — COMMENTED** [review]
May be better to add message "but received XX" and please capitalize the error message.

File: paddle/fluid/framework/executor.cc (hunk 7 above)
Comment (zhhsplendid) [reply paddle/fluid/framework/executor.cc]: Done

File: paddle/fluid/framework/executor.cc (hunk 8 above)
Comment (zhhsplendid) [reply paddle/fluid/framework/executor.cc]: Done, also changed a similar place at fetch

File: paddle/fluid/framework/executor.cc (hunk 9 above)
Comment (zhhsplendid) [reply paddle/fluid/framework/executor.cc]: Done

File: paddle/fluid/framework/parallel_executor.cc (hunk 10 above)
Comment (zhhsplendid) [reply paddle/fluid/framework/parallel_executor.cc]: Done

File: paddle/fluid/framework/parallel_executor.cc (hunk 11 above)
Comment (zhhsplendid) [reply paddle/fluid/framework/parallel_executor.cc]: Done

File: paddle/fluid/framework/parallel_executor.cc (hunk 12 above)
Comment (zhhsplendid) [reply paddle/fluid/framework/parallel_executor.cc]: Done

**Review by Aurelius84 — APPROVED** [review]

**Review by liym27 — APPROVED** [review]
LGTM

**Review by luotao1 — APPROVED** [review]

**Pull request merged**